### ISSUE TITLE
Create `legally` node.js API

### ISSUE THREAD

**Issue by michaelrambeau** [franciscop/legally#19]
Hello, we'd like to integrate data from `legally` in [Best of JavaScript](https://bestofjs.org/), an open-source application that gathers the trends about a curated list of more than 1000 projects related to the web platform (mainly JavaScript front-end and back-end).

Right now `legally` can be used from the command line but it would be great if it could be called programmatically from any node.js program, as if it was an API.

```js
const report = await legally('express') // to fetch JSON data about licenses
```

Maybe the core of the `legally` project could be extracted from the CLI application, a bit like Sindre Sorhus does with all his CLI projects, made of a core package and `*-cli` package.

We may refine the need after the first iterations, but it would be great if all command line parameters were available.

Thank you very much @franciscop for having *already* started to handle our request in this commit https://github.com/franciscop/legally/commit/09b9bd12d40a619327bd3ad800a5eb1656543ae4 you're so fast 😄 !

**Comment by michaelrambeau** [comment]
Hello @franciscop ,

Thank you, it works, here is an example of response we get, for `unstated` project:

```json
{
  "create-react-context@0.1.6": {
    "package": ["MIT"],
    "license": ["MIT"],
    "readme": []
  },
  "unstated@2.1.1": {
    "package": ["MIT"],
    "license": ["MIT"],
    "readme": []
  }
}
```

I wonder if it would be possible to get the report that shows the number and the percentage by type of license, like you do with the command line?

![image](https://user-images.githubusercontent.com/5546996/44518965-a97d3e80-a706-11e8-8988-9015d1d23607.png)

Maybe we could return this kind of structure:

```json
{
  "licenses": {
    "create-react-context@0.1.6": {
      "package": ["MIT"],
      "license": ["MIT"],
      "readme": []
    },
    "unstated@2.1.1": {
      "package": ["MIT"],
      "license": ["MIT"],
      "readme": []
    }
  },
  "report": {
    "total": 2,
    "MIT": 2
  }
}
```


I'm aware that I could do the calculation by myself, so don't worry if my request is troublesome ;)

**Comment by franciscop** [comment]
I'm working on making `fs-array` asynchronous, then I'd migrate `legally` to it. But I'm having some trouble with some advanced uses of `Proxy()` :'(

If I cannot get it to where I want by this weekend I'll just give up on this specific way and make a simpler solution.

About the second request, right now `lib/analyze.js` is too coupled to the representation. That bit is quite low priority on my list of things to do, so feel free to pick up some methods from there or do a PR. Also, ask away if there's any issue, the code quality/legibility of `legally` is quite bad.

**Comment by michaelrambeau** [comment]
Alright @franciscop , no problem at all, ignore my request, don't spoil your time, it's easy to `reduce` the results and show a report to users.

I will try to have a look at the code, but just to understand the context, when we call the node.js API, what does the code do with the File System?

I thought it was only about making HTTP requests in a recursive way to fetch data about a given package and the tree of all its dependencies, but I may miss something.

**Comment by franciscop** [comment]
Ah no, I didn't want to replicate npm's functionality of dependency resolution. So it creates a temporary folder `.tmp`, installs the package there and reads it and all the packages inside. Now that I'm looking into it, the installation is sync as well, so it'll block any live server so far.

This workflow is totally fine if you are running it in a single process with the CLI, but it doesn't allow for packages to be installed in parallel so it *will* block the main thread if the users can manually check them. READ: I highly recommend *not* using it yet for a live server.

You **will** also need to heavily sanitize the package names, since this is literally the command:

```js
'npm install ' + packages.join(' ') + ' --save --ignore-scripts'
```

See the source code of that bit: https://github.com/franciscop/legally/blob/master/lib/remote.js

Basically: this lib is made with the intention of being purely a CLI; the Node.js API is an afterthought and right now should be considered very unstable and vulnerable.

Potential issues so far:

- [ ] Blocking; not suitable for a live server
- [ ] Code injection; not suitable for a live server
- [ ] No analysis performed; needs refactoring
- [ ] Proper clean-up of temp files; needs investigation

**Comment by franciscop** [comment]
That said, I think it would be a good idea for the Node.js API to happen, so I will try to fix at least the actual security concerns.

**Comment by franciscop** [comment]
Step 1 done after much hard work: https://github.com/franciscop/magic-promises

Next I will make `fs-array` async based on this (easy).

**Comment by franciscop** [comment]
Security issue fixed. Migrated from `fs.{...}Sync()` to the package `fs-array`, which is fully async to avoid potential DDOS.

TODO: need to review if there's any issue with parallel installations as in a web server checking licenses. If so, would need to namespace those.

TODO: review the clean up script for `@scoped` packages. Might be buggy.

**Comment by franciscop** [comment]
Namespaced to avoid any potential issue with parallel installations.

**Comment by michaelrambeau** [comment]
Thank you very much Francisco! @franciscop 

I created a service hosted on Now.sh that uses `legally` node.js API to fetch data about licenses.

For a given package, it returns a summary by type of license ("MIT", "ISC")...

For example, for `unstated` package, calling the URL https://fetch-license.now.sh/unstated will return:

```json
{
  "status": "OK",
  "licenses": {
    "MIT": {
      "count": 2,
      "packages": ["create-react-context@0.1.6", "unstated@2.1.1"]
    },
    "ISC": {
      "count": 1,
      "packages": ["pack-unstated@1.0.0"]
    }
  }
}

```

Note: I only consider the "package" section of results returned by `legally`, I'm ignoring both `copying` and `readme` sections.

I'm working right now on integrating this service in _Best of JavaScript_.

**Comment by michaelrambeau** [comment]
Hello Franciso @franciscop 

We're getting closer of the release, let me share some feedback after a few tests I did on https://license-data--bestofjs.netlify.com

### Scoped packages

You already know it, but for scoped packages (such as `@cycle/core`), we get an error:

```
ENOENT: no such file or directory, lstat '/tmp/legally/pack-@cycle/core'
```
### Apache VS Apache 2.0?

I wonder if the license research could be improved.

Check `rxdb` package for example: https://license-data--bestofjs.netlify.com/projects/rxdb

I noticed that some packages have both the `Apache 2.0` license and the `Apache` license, like this package https://www.npmjs.com/package/aws-sign2.

Here is the response returned by `legally`

```json
{
  "aws-sign2@0.7.0": {
    "package": ["Apache 2.0"],
    "copying": ["Apache"],
    "readme": []
  }
}
```

Is it correct for you?
I think it should be `Apache 2.0` but it seems there is a problem when we parse the following file: https://github.com/request/aws-sign/blob/master/LICENSE

```
Apache License

Version 2.0, January 2004

http://www.apache.org/licenses/

TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION
``` 

Since the text contains "Apache License", it's considered as just "Apache" instead of "Apache 2.0"

I found a similar problem in an other package (`atob`) when we parse the following content:

```
At your option you may choose either of the following licenses:

  * The MIT License (MIT)
  * The Apache License 2.0 (Apache-2.0)


The MIT License (MIT)

Copyright (c) 2015 AJ ONeal

Permission is hereby granted...

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", ....


                                 Apache License
                           Version 2.0, January 2004
                        http://www.apache.org/licenses/
```

Since the string `Apache License` is found at the end, it's considered as `Apache`.

Maybe the solution would be to search *first* for `Apache 2.0` license and then, *if* it's not found, search for just "Apache"?

**Comment by michaelrambeau** [comment]
I found out that the scoped packages issue has been corrected by this fix https://github.com/franciscop/legally/commit/928af7f6d8f6d96f89839e65b93444cf1d65cd66.

About the Apache license issue I mentioned above, as a simple fix from my side, I'm ignoring "Apache" licenses if an "Apache 2.0" license has been found, for a given package (done in this commit: https://github.com/michaelrambeau/bestofjs-now-microservices/commit/d82195217c108b474c72b14338aebec9d6e22588#diff-8b758f79f0652672e3dd40e786a0379dR34)

**Comment by franciscop** [comment]
I think this (including the Apache 2.0 issue) must have been fixed now. Please let me know when you try it, and thanks for all these reports!

**Comment by franciscop** [comment]
Found an interesting thing: https://www.npmjs.com/package/validate-npm-package-name

**Comment by michaelrambeau** [comment]
Thank you Francisco, I've just updated the API code, including `validate-npm-package-name` to check the inputs and upgrading `legally` to the latest version.
I'm going to run more tests today.
